### ISSUE TITLE
[Feature Request] Add support for the ADB deployment workflow and HTTP trigger hooks to the built-in updater.

### ISSUE THREAD

**Issue by oyugio** [reisxd/TizenTubeCobalt#314]
Google is enforcing strict side-loading restrictions and a mandatory 24-hour verification delay on standard Android Package Installers. The internal TV-based updater (TizenTube Cobalt Güncelleyici) will soon face severe OS-level hurdles in non-rooted Android TV environments.

Since TV environments lack the proper background authorization for Shizuku, the most reliable deployment method is remote execution via a local network ADB bridge (adb install -r).

### Proposal:

1. Updater UI Toggle: Add an option under Updater Settings: Disable Native Package Installer (ADB/Manual Mode).

2. Local APK Export
When this option is enabled, the updater will download the release APK directly to local storage (e.g., /sdcard/Download/TizenTubeCobalt-release.apk) and will completely suppress the PackageInstaller Intent prompt.

3. HTTP Trigger Hook: Expose a minimal local background HTTP endpoint (e.g., GET /status) to indicate download completion. This allows LAN-side automation scripts to trigger an automated adb install -r flow.

**Comment by oyugio** [comment]
Here is a minimal Proof of Concept (PoC) implementation for the proposed C++ HTTP trigger hook and the LAN-side Python ADB deployment script:

### 1. C++ POSIX Socket Trigger Hook (`cobalt/browser/application.cc`)
```cpp
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <thread>

void StartHttpTriggerServer(int port) {
    std::thread([port]() {
        int server_fd = socket(AF_INET, SOCK_STREAM, 0);
        int opt = 1;
        setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));
        sockaddr_in address{AF_INET, htons(port), INADDR_ANY};
        
        bind(server_fd, (struct sockaddr*)&address, sizeof(address));
        listen(server_fd, 3);
        
        while (true) {
            int new_socket = accept(server_fd, nullptr, nullptr);
            if (new_socket >= 0) {
                const char* response = "HTTP/1.1 200 OK\r\nContent-Length: 16\r\n\r\nREADY_FOR_UPDATE";
                send(new_socket, response, strlen(response), 0);
                close(new_socket);
            }
        }
    }).detach();
}







import requests
import subprocess
import os

TV_IP = "192.168.1.100"  # Android TV LAN IP address
TRIGGER_PORT = 8080
APK_PATH = "./out/android-arm_release/cobalt.apk"

def deploy():
    if not os.path.exists(APK_PATH):
        print("[-] APK file not found. Check build output.")
        return

    try:
        res = requests.get(f"http://{TV_IP}:{TRIGGER_PORT}", timeout=3)
        if "READY_FOR_UPDATE" in res.text:
            print("[+] TV trigger responded. Establishing ADB connection...")
            subprocess.run(["adb", "connect", f"{TV_IP}:5555"], check=True)
            print("[+] Starting silent installation...")
            subprocess.run(["adb", "-s", f"{TV_IP}:5555", "install", "-r", "-d", APK_PATH], check=True)
            print("[+] Update completed successfully!")
    except Exception as e:
        print(f"[-] Deployment error: {e}")

if __name__ == "__main__":
    deploy()

**Comment by oyugio** [comment]
### Refined PoC (Addressing Socket CPU Spin & Execution Context)

Based on further code analysis, here is the optimized PoC preventing potential CPU hogging on Android TV and improving ADB deployment error handling on the host PC:

#### 1. Optimized C++ Socket Server (`cobalt/browser/application.cc`)
*Fixes infinite CPU loop on `accept()` socket errors and ensures proper socket cleanups.*

```cpp
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <thread>
#include <chrono>

void StartHttpTriggerServer(int port) {
    std::thread([port]() {
        int server_fd = socket(AF_INET, SOCK_STREAM, 0);
        if (server_fd < 0) return;

        int opt = 1;
        setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

        sockaddr_in address{};
        address.sin_family = AF_INET;
        address.sin_addr.s_addr = INADDR_ANY;
        address.sin_port = htons(port);

        if (bind(server_fd, (struct sockaddr*)&address, sizeof(address)) < 0) {
            close(server_fd);
            return;
        }

        if (listen(server_fd, 3) < 0) {
            close(server_fd);
            return;
        }

        while (true) {
            int new_socket = accept(server_fd, nullptr, nullptr);
            if (new_socket < 0) {
                // Prevent 100% CPU spin lock on socket accept errors
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
                continue;
            }

            const char* response = "HTTP/1.1 200 OK\r\nContent-Length: 16\r\n\r\nREADY_FOR_UPDATE";
            send(new_socket, response, strlen(response), 0);
            close(new_socket);
        }
        close(server_fd);
    }).detach();
}









import requests
import subprocess
import os

TV_IP = "192.168.1.100"  # Android TV LAN IP address
TRIGGER_PORT = 8080
APK_PATH = "./out/android-arm_release/cobalt.apk"
PACKAGE_NAME = "com.tizentube.cobalt"  # Target application package

def deploy():
    if not os.path.exists(APK_PATH):
        print(f"[-] APK file not found at: {APK_PATH}")
        return

    # 1. Check TV HTTP Endpoint Status
    try:
        res = requests.get(f"http://{TV_IP}:{TRIGGER_PORT}", timeout=3)
        if "READY_FOR_UPDATE" not in res.text:
            print("[-] TV trigger responded, but status is not ready.")
            return
        print("[+] TV trigger verified.")
    except requests.exceptions.RequestException as e:
        print(f"[-] Connection to TV trigger failed: {e}")
        return

    # 2. Establish ADB Connection
    try:
        print("[+] Connecting ADB...")
        subprocess.run(["adb", "connect", f"{TV_IP}:5555"], check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as e:
        print(f"[-] ADB Connect Failed: {e.stderr}")
        return

    # 3. Perform Silent Installation & Re-launch
    try:
        print("[+] Installing APK silently (-r -d)...")
        subprocess.run(
            ["adb", "-s", f"{TV_IP}:5555", "install", "-r", "-d", APK_PATH],
            check=True, capture_output=True, text=True
        )
        print("[+] Update installed successfully!")
        
        # Optional: Auto-launch the app post-installation
        subprocess.run(["adb", "-s", f"{TV_IP}:5555", "shell", "monkey", "-p", PACKAGE_NAME, "-c", "android.intent.category.LAUNCHER", "1"], capture_output=True)
        print("[+] Application restarted on TV.")
    except subprocess.CalledProcessError as e:
        print(f"[-] ADB Installation Failed: {e.stderr}")

if __name__ == "__main__":
    deploy()

**Comment by reisxd** [comment]
There isn't a good or a proper way to implement this.

**Comment by oyugio** [comment]
Understood! Thanks for checking and clarifying.
